File: javascript/17-taxonomy-trap/test/vehicle.test.js
import assert from 'node:assert';
import test from 'node:test';
import { Car, Truck, ElectricCar, DieselCar, ElectricTruck, DieselTruck } from '../src/vehicle.js';

test('car daily rate', () => {
  const v = new Car('Toyota', 'Camry', 2020);
  assert.strictEqual(v.daily_rate(), 40);
});

test('truck daily rate', () => {
  const v = new Truck('Ford', 'F-150', 2020);
  assert.strictEqual(v.daily_rate(), 80);
});

test('electric car fuel cost', () => {
  const v = new ElectricCar('Tesla', 'Model 3', 2020);
  assert.strictEqual(v.fuel_cost(5), 0);
});

test('diesel car fuel cost', () => {
  const v = new DieselCar('VW', 'Jetta', 2020);
  assert.strictEqual(v.fuel_cost(5), 25);
});

test('electric truck fuel cost', () => {
  const v = new ElectricTruck('Rivian', 'R1T', 2020);
  assert.strictEqual(v.fuel_cost(5), 0);
});

test('diesel truck fuel cost', () => {
  const v = new DieselTruck('Ford', 'F-250', 2020);
  assert.strictEqual(v.fuel_cost(5), 75);
});

test('car stores brand', () => {
  const v = new Car('Toyota', 'Camry', 2020);
  assert.strictEqual(v.brand, 'Toyota');
});

test('truck stores model', () => {
  const v = new Truck('Ford', 'F-150', 2020);
  assert.strictEqual(v.model, 'F-150');
});

File: csharp/17-taxonomy-trap/test/vehicle_test.cs
using Xunit;

public class VehicleTest
{
    [Fact]
    public void CarDailyRate()
    {
        var v = new Car("Toyota", "Camry", 2020);
        Assert.Equal(40, v.daily_rate());
    }

    [Fact]
    public void TruckDailyRate()
    {
        var v = new Truck("Ford", "F-150", 2020);
        Assert.Equal(80, v.daily_rate());
    }

    [Fact]
    public void ElectricCarFuelCost()
    {
        var v = new ElectricCar("Tesla", "Model 3", 2020);
        Assert.Equal(0, v.fuel_cost(5));
    }

    [Fact]
    public void DieselCarFuelCost()
    {
        var v = new DieselCar("VW", "Jetta", 2020);
        Assert.Equal(25, v.fuel_cost(5));
    }

    [Fact]
    public void ElectricTruckFuelCost()
    {
        var v = new ElectricTruck("Rivian", "R1T", 2020);
        Assert.Equal(0, v.fuel_cost(5));
    }

    [Fact]
    public void DieselTruckFuelCost()
    {
        var v = new DieselTruck("Ford", "F-250", 2020);
        Assert.Equal(75, v.fuel_cost(5));
    }

    [Fact]
    public void CarStoresBrand()
    {
        var v = new Car("Toyota", "Camry", 2020);
        Assert.Equal("Toyota", v.Brand);
    }

    [Fact]
    public void TruckStoresModel()
    {
        var v = new Truck("Ford", "F-150", 2020);
        Assert.Equal("F-150", v.Model);
    }

    [Fact]
    public void DieselCarRentalTotalCombinesRateFuelInsuranceAndGps()
    {
        var v = new DieselCar("VW", "Jetta", 2020);
        Assert.Equal(195, v.rental_total(3, true));
    }

    [Fact]
    public void ElectricTruckRentalTotalCombinesTruckRateAndInsurance()
    {
        var v = new ElectricTruck("Rivian", "R1T", 2020);
        Assert.Equal(200, v.rental_total(2, false));
    }

    [Fact]
    public void DieselTruckRentalTotalIncludesHigherFuelCost()
    {
        var v = new DieselTruck("Ford", "F-250", 2020);
        Assert.Equal(246, v.rental_total(2, true));
    }
}
